internal/core/command/device.go

```diff
@@ -16,28 +16,45 @@ package command
 
 import (
 	"context"
+	"fmt"
 	"net/http"
+	"strings"
+
+	"github.com/edgexfoundry/edgex-go/internal"
+	"github.com/edgexfoundry/edgex-go/internal/core/command/config"
+	"github.com/edgexfoundry/edgex-go/internal/core/command/errors"
+	"github.com/edgexfoundry/edgex-go/internal/core/command/interfaces"
 
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/metadata"
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
 	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
 
-	"github.com/edgexfoundry/edgex-go/internal/core/command/config"
-	"github.com/edgexfoundry/edgex-go/internal/core/command/errors"
-	"github.com/edgexfoundry/edgex-go/internal/core/command/interfaces"
+const (
+	deviceIdPositionInUrlPath  int = 4
+	commandIdPositionInUrlPath int = 6
+	wellFormedUrlPathLength    int = 7
 )
 
 func executeCommandByDeviceID(
-	ctx context.Context,
-	deviceID string,
-	commandID string,
+	r *http.Request,
 	body string,
-	queryParams string,
-	isPutCommand bool,
 	lc logger.LoggingClient,
 	dbClient interfaces.DBClient,
-	deviceClient metadata.DeviceClient) (string, error) {
+	deviceClient metadata.DeviceClient,
+	httpCaller internal.HttpCaller) (string, error) {
+
+	var deviceID string
+	var commandID string
+	var ctx context.Context
+
+	if r != nil {
+		ctx = r.Context()
+		deviceID, commandID = extractDeviceIdAndCommandIdFromRequest(r)
+	} else {
+		return "", errors.NewErrExtractingInfoFromRequest(commandID, deviceID)
+	}
```

Comment:
Can we invert this conditional logic? This would make this simpler to read, and eliminate the need for side effects. This stood out more so since there seems to me a not so obvious bug with the `commandID` and `deviceID` within the `else` portion of the conditional. Those variables do not seem to be set and will always be empty strings.

By inverting the logic we would essentially be eliminating the `else` portion. We can check if the request is `nil` and if it is we can return the error. This will allow us to remove the declarations (`var`) and instead use `:=` which eliminates side effects and by proxy bugs.